File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/service/IntentUtilsTest.java
package com.emarsys.mobileengage.service;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;

import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.TimeoutUtils;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.HashMap;
import java.util.Map;

import static junit.framework.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class IntentUtilsTest {

    static {
        mock(Activity.class);
        mock(PackageManager.class);
    }

    private Context context;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void init() {
        context = InstrumentationRegistry.getTargetContext();

    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateLaunchIntent_remoteIntentMustNotBeNull() {
        IntentUtils.createLaunchIntent(null, context);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateLaunchIntent_contextMustNotBeNull() {
        Intent intent = new Intent();
        intent.putExtra("key", "value");
        IntentUtils.createLaunchIntent(intent, null);
    }

    @Test
    public void testCreateLaunchIntent() {
        Intent launchIntentForPackage = new Intent();
        PackageManager pm = mock(PackageManager.class);
        when(pm.getLaunchIntentForPackage(anyString())).thenReturn(launchIntentForPackage);
        Activity mockActivity = mock(Activity.class);
        when(mockActivity.getPackageManager()).thenReturn(pm);
        when(mockActivity.getPackageName()).thenReturn("packageName");


        Intent intent = new Intent();
        Bundle bundle = new Bundle();
        bundle.putString("key", "value");
        intent.putExtras(bundle);
        Intent result = IntentUtils.createLaunchIntent(intent, mockActivity);

        Assert.assertTrue(launchIntentForPackage == result);

        Bundle launcherBundle = launchIntentForPackage.getExtras();

        Assert.assertEquals(bundle.keySet(), launcherBundle.keySet());
        for (String key : bundle.keySet()) {
            Assert.assertEquals(bundle.get(key), launcherBundle.get(key));
        }
    }

    @Test
    public void testCreateLaunchIntent_withNoBundleInIntent() {
        Intent launchIntentForPackage = new Intent();
        PackageManager pm = mock(PackageManager.class);
        when(pm.getLaunchIntentForPackage(anyString())).thenReturn(launchIntentForPackage);
        Activity mockActivity = mock(Activity.class);
        when(mockActivity.getPackageManager()).thenReturn(pm);
        when(mockActivity.getPackageName()).thenReturn("packageName");

        Intent result = IntentUtils.createLaunchIntent(new Intent(), mockActivity);

        Assert.assertTrue(launchIntentForPackage == result);
        Assert.assertEquals(null, launchIntentForPackage.getExtras());
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateNotificationHandlerServiceIntent_remoteMessageDataMustNotBeNull() {
        IntentUtils.createNotificationHandlerServiceIntent(context, null, 0, "action");
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateNotificationHandlerServiceIntent_contextMustNotBeNull() {
        IntentUtils.createNotificationHandlerServiceIntent(null, new HashMap<String, String>(), 0, "action");
    }

    @Test
    public void createNotificationHandlerServiceIntent() {
        int notificationId = 987;

        Map<String, String> remoteMessageData = new HashMap<>();
        remoteMessageData.put("key1", "value1");
        remoteMessageData.put("key2", "value2");

        Intent resultIntent = IntentUtils.createNotificationHandlerServiceIntent(context, remoteMessageData, notificationId, "action");
        assertEquals("action", resultIntent.getAction());
        Bundle payload = resultIntent.getBundleExtra("payload");
        assertEquals("value1", payload.getString("key1"));
        assertEquals("value2", payload.getString("key2"));
        assertEquals(notificationId, payload.getInt("notification_id"));
    }

    @Test
    public void createNotificationHandlerServiceIntent_withoutAction() {
        Intent resultIntent = IntentUtils.createNotificationHandlerServiceIntent(context, new HashMap<String, String>(), 0, null);
        assertEquals(null, resultIntent.getAction());
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateNotificationHandlerServicePendingIntent_remoteMessageDataMustNotBeNull() {
        IntentUtils.createNotificationHandlerServicePendingIntent(context, null, 0, "action");
    }

    @Test(expected = IllegalArgumentException.class)
    public void testCreateNotificationHandlerServicePendingIntent_contextMustNotBeNull() {
        IntentUtils.createNotificationHandlerServicePendingIntent(null, new HashMap<String, String>(), 0, "action");
    }

}

File: mobile-engage/src/androidTest/java/com/emarsys/mobileengage/service/IntentUtilsTest.kt
package com.emarsys.mobileengage.service

import android.app.Activity
import android.app.PendingIntent
import android.app.PendingIntent.FLAG_IMMUTABLE
import android.app.PendingIntent.FLAG_UPDATE_CURRENT
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import com.emarsys.core.util.AndroidVersionUtils.isBelowUpsideDownCake
import com.emarsys.core.util.AndroidVersionUtils.isUpsideDownCakeOrHigher
import com.emarsys.mobileengage.di.setupMobileEngageComponent
import com.emarsys.mobileengage.fake.FakeMobileEngageDependencyContainer
import com.emarsys.mobileengage.service.IntentUtils.createLaunchPendingIntent
import com.emarsys.mobileengage.service.IntentUtils.createNotificationHandlerServiceIntent
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkStatic
import io.mockk.slot
import io.mockk.unmockkStatic
import io.mockk.verify
import org.junit.After
import org.junit.Before
import org.junit.Test

class IntentUtilsTest  {
    private companion object {
        const val BACKGROUND_ACTIVITY_START_MODE_KEY = "android.activity.pendingIntentCreatorBackgroundActivityStartMode"
        const val TITLE = "title"
        const val BODY = "body"
        const val CHANNEL_ID = "channelId"
        const val COLLAPSE_ID = "testCollapseId"
        const val MULTICHANNEL_ID = "test multiChannel id"
        const val SID = "test sid"
        const val SMALL_RESOURCE_ID = 123
        const val COLOR_RESOURCE_ID = 456
        val notificationData = NotificationData(
            null,
            null,
            null,
            TITLE,
            BODY,
            CHANNEL_ID,
            campaignId = MULTICHANNEL_ID,
            sid = SID,
            smallIconResourceId = SMALL_RESOURCE_ID,
            colorResourceId = COLOR_RESOURCE_ID,
            collapseId = COLLAPSE_ID,
            operation = NotificationOperation.INIT.name,
            actions = null,
            inapp = null,
            u = "{\"customField\":\"customValue\"}",
            message_id = "messageId"
        )
    }

    private lateinit var context: Context
    private lateinit var launchIntentForPackage: Intent
    private lateinit var mockPackageManager: PackageManager
    private lateinit var mockActivity: Activity
    private val bundleCaptor = slot<Bundle>()
    private val intentCaptor = slot<Intent>()

    @Before
    fun init() {
        setupMobileEngageComponent(FakeMobileEngageDependencyContainer())
        context = getTargetContext()
        mockkStatic(PendingIntent::class)

        launchIntentForPackage = Intent()
        mockPackageManager = mockk(relaxed = true)
        every { mockPackageManager.getLaunchIntentForPackage(any()) } returns launchIntentForPackage

        mockActivity = mockk(relaxed = true)
        every { mockActivity.packageManager } returns mockPackageManager
        every { mockActivity.packageName } returns "com.emarsys.mobileengage.test"
        every {
            PendingIntent.getActivity(
                mockActivity,
                0,
                launchIntentForPackage,
                any(),
                any()
            )
        } returns mockk()
    }

    @After
    fun tearDown() {
        unmockkStatic(PendingIntent::class)
        bundleCaptor.clear()
        intentCaptor.clear()
    }

    @Test
    fun testCreateLaunchIntent() {
        if(isBelowUpsideDownCake) {
            val remoteIntent = Intent()
            val expectedExtras = Bundle()
            expectedExtras.putString("key", "value")
            remoteIntent.putExtras(expectedExtras)

            createLaunchPendingIntent(remoteIntent, mockActivity)

            verify {
                PendingIntent.getActivity(
                    mockActivity,
                    0,
                    capture(intentCaptor),
                    FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE,
                    null
                )
            }

            intentCaptor.captured.extras?.getString("key") shouldBe "value"
        }
    }

    @Test
    fun testCreateLaunchIntent_shouldCreatePendingIntentWithApplicationOptionsAboveApiLevel34() {
        if(isUpsideDownCakeOrHigher) {
            val remoteIntent = Intent()
            val expectedExtras = Bundle()
            expectedExtras.putString("key", "value")
            remoteIntent.putExtras(expectedExtras)

            createLaunchPendingIntent(remoteIntent, mockActivity)

            verify {
                PendingIntent.getActivity(
                    mockActivity,
                    0,
                    capture(intentCaptor),
                    FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE,
                    capture(bundleCaptor)
                )
            }

            bundleCaptor.captured.getInt(BACKGROUND_ACTIVITY_START_MODE_KEY) shouldBe 1
            intentCaptor.captured.extras?.getString("key") shouldBe "value"
        }
    }

    @Test
    fun testCreateLaunchIntent_withNoBundleInIntent() {
        if (isBelowUpsideDownCake) {
            createLaunchPendingIntent(Intent(), mockActivity)
            launchIntentForPackage.extras shouldBe null

            verify {
                PendingIntent.getActivity(
                    mockActivity,
                    0,
                    capture(intentCaptor),
                    FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE,
                    null
                )
            }

            intentCaptor.captured.extras shouldBe null
        }
    }

    @Test
    fun testCreateLaunchIntent_withNoBundleInIntent_aboveApiLevel34() {
        if (isUpsideDownCakeOrHigher) {
            createLaunchPendingIntent(Intent(), mockActivity)
            launchIntentForPackage.extras shouldBe null

            verify {
                PendingIntent.getActivity(
                    mockActivity,
                    0,
                    capture(intentCaptor),
                    FLAG_UPDATE_CURRENT or FLAG_IMMUTABLE,
                    capture(bundleCaptor)
                )
            }

            bundleCaptor.captured.getInt(BACKGROUND_ACTIVITY_START_MODE_KEY) shouldBe 1
            intentCaptor.captured.extras shouldBe null
        }
    }

    @Test
    fun testCreateLaunchIntent_whenLaunchIntentForPackageIsNull() {
        every { mockPackageManager.getLaunchIntentForPackage(any()) } returns null
        val result = createLaunchPendingIntent(Intent(), mockActivity)
        result shouldBe null
    }

    @Test
    fun testCreateNotificationHandlerServiceIntent() {
        val resultIntent = createNotificationHandlerServiceIntent(
            context,
            notificationData,
            "action"
        )
        resultIntent.action shouldBe "action"
        val payload = resultIntent.getParcelableExtra<NotificationData>("payload")
        payload shouldBe notificationData
    }

    @Test
    fun createNotificationHandlerServiceIntent_withoutAction() {
        val resultIntent = createNotificationHandlerServiceIntent(
            context,
            notificationData,
            null
        )
        resultIntent.action shouldBe null
    }
}